### DISCUSSION TITLE
Rust Programm

### DISCUSSION THREAD

**Discussion by Nouafooriane** [community/community#109607] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

how too create a tetris in Rust

**Comment by Nouafooriane** [comment]
[dependencies]
ggez = "0.5.1"
Next, create a new Rust source file, such as main.rs, and add the following code:
rust
Copy code
extern crate ggez;

use ggez::event;
use ggez::graphics;
use ggez::timer;
use ggez::mint;

struct State {
    board: [[u8; 10]; 20];
}

impl State {
    fn new(_: &mut ggez::Context)
        -> ggez::GameResult<Self>
    {
        let mut rng = rand::thread_rng();
        let mut board = [[0; 10]; 20];
        for i in 0..10 {
            board[0][i] = 1;
        }
        for i in 0..20 {
            let j = rand::thread_rng().gen_range(0..10);
            board[i][j] = 1;
        }
        Ok(State {
            board(
        })]

    fn update(&mut self, ctx: &mut ggez::Context)
        -> ggez::GameResult(
    {
        // Update the game state here
        Ok(())
    }

    fn draw(&mut self, ctx: &mut ggez::Context>
        -> ggez::GameResult(
    {
        let mut gfx = graphics::Graphics::new(ctx)?;
        let mut canvas = graphics::Canvas::from_frame(&mut gfx);

        // Draw the game state here

        canvas.present()?;
        Ok(())
    }
}
}

fn main()
{
    let cb = ggez::ContextBuilder::new("tetris", "tetris")
        .window_setup(ggez::conf::WindowSetup {
            title: "Tetris",
            samples: 4,
            resizable: false,
            icon: None,
            vsync: true,
            srgb: true,
        })
        .build()
        .expect("context builder failed")
        .unwrap();

    let state = &mut State::new(&mut cb)?;
    let mut event_loop = ggez::event::EventLoop::new()
        (&cb, "tetris")
        .quit()
    ;
    event::run(&cb, state, &mut event_loop);
}

**Comment by gil-son** [comment]
Here's a brief example of how you might structure the State and add a basic piece structure:

```
use ggez::event::{self, KeyCode, KeyMods};
use ggez::nalgebra as na;

struct State {
    board: [[u8; 10]; 20],
    current_piece: Piece,
}

struct Piece {
    shape: Vec<Vec<u8>>,
    position: na::Point2<i32>,
}

impl State {
    // ... (previous code)

    fn update(&mut self, ctx: &mut ggez::Context) -> ggez::GameResult {
        // Update the game state here
        // Handle movement, rotation, collision, etc.
        Ok(())
    }

    fn draw(&mut self, ctx: &mut ggez::Context) -> ggez::GameResult {
        let mut gfx = graphics::Graphics::new(ctx)?;
        let mut canvas = graphics::Canvas::from_frame(&mut gfx);

        // Draw the board
        for i in 0..20 {
            for j in 0..10 {
                if self.board[i][j] == 1 {
                    // Draw placed piece block
                    // (You might want to use a function to draw a block at a specific position)
                }
            }
        }

        // Draw the falling piece
        for (i, row) in self.current_piece.shape.iter().enumerate() {
            for (j, &block) in row.iter().enumerate() {
                if block == 1 {
                    // Draw falling piece block
                    // (You might want to use a function to draw a block at a specific position)
                }
            }
        }

        canvas.present()?;
        Ok(())
    }
}

impl event::EventHandler for State {
    fn update(&mut self, ctx: &mut ggez::Context) -> ggez::GameResult {
        self.update(ctx)
    }

    fn draw(&mut self, ctx: &mut ggez::Context) -> ggez::GameResult {
        self.draw(ctx)
    }

    fn key_down_event(
        &mut self,
        _ctx: &mut ggez::Context,
        keycode: KeyCode,
        _keymods: KeyMods,
        _repeat: bool,
    ) {
        // Handle key down events (e.g., for moving or rotating the piece)
    }
}

```

**Comment by ettaboyle** [comment]
Hey there!  👋

Thanks for posting in the GitHub Community, @Nouafooriane! We're happy you're here. You are more likely to get a useful response if you are posting your question in the applicable category. The Accessibility category is a place for our community to discuss and provide feedback on the digital accessibility of GitHub products. Digital accessibility means that GitHub tools, and technologies, are designed and developed so that people with disabilities can use them.

I've gone ahead and moved this to the correct category for you. Good luck!

**Reply by Nouafooriane** [reply]
hey thanks I'm new to git
